src/vm/types/mod.rs

```diff
@@ -0,0 +1,450 @@
+pub mod serialization;
+mod signatures;
+
+use std::hash::{Hash, Hasher};
+use std::{fmt, cmp};
+use std::convert::TryInto;
+use std::collections::BTreeMap;
+
+use address::c32;
+use vm::representations::{ClarityName, ContractName, SymbolicExpression, SymbolicExpressionType};
+use vm::errors::{RuntimeErrorType, UncheckedError, InterpreterResult as Result, IncomparableError, InterpreterError};
+use util::hash;
+
+pub use vm::types::signatures::{
+    AtomTypeIdentifier, TupleTypeSignature, AssetIdentifier,
+    TypeSignature, FunctionType, ListTypeData, FunctionArg, parse_name_type_pairs};
+
+pub const MAX_VALUE_SIZE: i128 = 1024 * 1024; // 1MB
+
+#[derive(Debug, Clone, Eq, Serialize, Deserialize)]
+pub struct TupleData {
+    type_signature: TupleTypeSignature,
+    pub data_map: BTreeMap<ClarityName, Value>
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
+pub struct BuffData {
+    pub data: Vec<u8>,
+}
+
+#[derive(Debug, Clone, Eq, Serialize, Deserialize)]
+pub struct ListData {
+    pub data: Vec<Value>,
+    pub type_signature: ListTypeData
+}
+
+// a standard principal is a version byte + hash160 (20 bytes)
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Serialize, Deserialize)]
+pub struct StandardPrincipalData(pub u8, pub [u8; 20]);
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Serialize, Deserialize)]
+pub enum PrincipalData {
+    StandardPrincipal(StandardPrincipalData),
+    ContractPrincipal(ContractName),
+    QualifiedContractPrincipal { sender: StandardPrincipalData,
+                                 name: ContractName },
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Hash)]
+pub struct OptionalData {
+    pub data: Option<Box<Value>>,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Hash)]
+pub struct ResponseData {
+    pub committed: bool,
+    pub data: Box<Value>,
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Serialize, Deserialize)]
+pub enum Value {
+    Int(i128),
+    Bool(bool),
+    Buffer(BuffData),
+    List(ListData),
+    Principal(PrincipalData),
+    Tuple(TupleData),
+    Optional(OptionalData),
+    Response(ResponseData)
+}
+
+define_named_enum!(BlockInfoProperty {
+    Time("time"),
+    VrfSeed("vrf-seed"),
+    HeaderHash("header-hash"),
+    BurnchainHeaderHash("burnchain-header-hash"),
```

Comment:
Don't we also need the block height and the parent header hash?

Reply:
`block-height` is a native global, rather than an argument to the `block-info` function. It's defined in the `variables.rs` file.